### PR TITLE
NVDAObjects/UIA: recognize UWP tooltips

### LINKED ISSUE #8118: Tool Tips are not read in UWP apps when "report tooltips" is enabled
If "Report Tooltips" is turned on, tooltip messages are reported in Desktop programs but not for UWP apps.

Steps to Reproduce:
1. Turn on report tooltips in object presentation.
2. Open a modern program such as Microsoft Edge.
3. Tab through the UI elements, not a web page. This is the area that includes the address bar, hub, etc.
Expected result: Tool tips should be read for most controls.
Actual Results: No tool tips are read.
But if you open Narrator, tool tips are read, and this is important because the tool tips contain the access keys needed to get to menus and other controls quickly. Without this, NVDA users will have no easy way to learn the access keys in this and other situations.
This is on Windows 10 build 17127.1 on a recent Master snapshot.

Thank you very much for your time and I look forward to any fixes you may be able to provide.

Sincerely,

Brandon Tyson

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #8118 

### Summary of the issue:
Recognize UWP/XAML tooltips such as ones from Microsoft Edge as proper tooltips.

### Description of how this pull request fixes the issue:
Tooltips from universal apps are powered by XAML and has a specific class name (for now). Prior to this, these tooltips were not announced by NVDA. Cases include Edge's Settings button tooltip, which displays hotkey information (Alt+X) in Insider Preview builds. Thus recognize them as proper tooltips through an overlay class powered by NVDAObjects.behaviors.ToolTip.

### Testing performed:
Tested Edge on Windows 10 Versions 1703 (Creators Update), 1709 (Fall Creators Update) and 1803 preview (Spring Creators Update).

### Known issues with pull request:
None at this time.

### Change log entry:
Bug fixes: In Windows 10, NVDA will announce tooltips from universal apps if NVDA is configured to report tooltips in object presentation dialog.

### CODE REVIEW

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [25 mrt. 2018 19:44 CEST](https://github.com/nvaccess/nvda/pull/8124#issue-177301490 "2018-03-25T17:44:03Z - Replied by Github Reply Comments"):
> Tooltips from universal apps are powered by XAML and has a specific class name (for now).

What UIA control type do these have? It would be nice if we could also support WPF tooltips this way, though I have no idea where to find these for testing purposes.

**Comment by josephsl** [comment]
Hi, haven’t investigated that far, but I assume they’re regular tooltips.

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [26 mrt. 2018 09:21 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376069890 "2018-03-26T07:21:49Z - Replied by Github Reply Comments"):
> Hi,
> 
> The tooltips are recognized as tooltips. Thanks.

In that case, why not just treat every UIA tooltip as tooltip?

**Comment by josephsl** [comment]
Hi, not yet, as one of them is volume slider tooltip, which is annoying. Thanks.

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [26 mrt. 2018 16:48 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376192606 "2018-03-26T14:48:28Z - Replied by Github Reply Comments"):
> Hi, not yet, as one of them is volume slider tooltip, which is annoying. Thanks.

Could you please elaborate?

**Comment by josephsl** [comment]
Hi, when you change audio volume, a tooltip comes up and fires tooltip open event. Although the framework is different, it is still annoying. Thus, I’ve intentionally limited tooltip announcement to XAML tooltips for now. Thanks.

**Comment by leonardder** [comment]
[**@josephsl**](https://github.com/josephsl) commented on [26 Mar 2018, 17:44 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-376212903 "2018-03-26T15:44:31Z - Replied by Github Reply Comments"):
> Hi, when you change audio volume, a tooltip comes up and fires tooltip open event. Although the framework is different, it is still annoying.

I think many people consider tool tips to be annoying. I prefer consistency over subjective decisions about what is considered annoying and what is not. @MichaeldCurran: Thoughts?

**Comment by michaelDCurran** [comment]
I tend to agree with @leonardder.  If a choose to  have tooltips reported, then I want to hear all tooltips. Obviously this can get annoying, thus why it is disabled by default in NVDA.

**Comment by josephsl** [comment]
Hi,

Testing shows that I had incomplete assumptions: tooltip for volume slider is fired infrequently, thus I'll follow above recommendations.

Thanks.

**Comment by josephsl** [comment]
Hi,

Note that new UIA tool tip handling works properly on Windows 8 and later. An additional benefit is that NVDA will now also announce volume changes (if that is announced, that is).

Thanks.

**Review by michaelDCurran — APPROVED** [review]

**Comment by Robert-J-H** [comment]
Hi
I must say that the Volume Level tooltip is very distracting.
Although it closes essentially #5726, the opinions uttered there aren't by no way unanimous.

It is especially annoying if you adjust the volume during say all.
It might happen that it is announced much later (after e.g. a paragraph).

It is not very user-friendly if the possible decision for tool tips is
"Hav'em all or none".
It is a special case tool tip because:
1. It gives the immediate result for a user input 
2. The effect can also be heard by other means (playing audio, running speech)
As a musician, I clearly prefer the second way.
More so as the user intentionally lowers or rises the volume because of what she/he heard or didn't hear and not because the numeric value wasn't to her/his liking.
In other words, the cause/effect chain is broken or at least forked into two outputs. 
It would be great to have this tool tip only announced for when the audio device is idle.
Does this make sense? ;-)

**Comment by leonardder** [comment]
[**@Robert-J-H**](https://github.com/Robert-J-H) commented on [21 Apr 2018, 08:45 CEST](https://github.com/nvaccess/nvda/pull/8124#issuecomment-383272454 "2018-04-21T06:45:03Z - Replied by Github Reply Comments"):
> It is especially annoying if you adjust the volume during say all.
> It might happen that it is announced much later (after e.g. a paragraph).

May be this is more a general problem of reporting tooltips during say all, not specifically to the volume tool tip.

> It is not very user-friendly if the possible decision for tool tips is
> "Hav'em all or none".

What do you suggest otherwise? I don't think it is user friendly either to distinguish between several tool tips the user does not know about.

**Comment by josephsl** [comment]
Hi, this is due to the fact that the object in question fires tool tip opened event infrequently. I had similar objections due to this being not quite consistent, but ultimately it is up to users to decide. Thanks.

**Comment by Robert-J-H** [comment]
Yes, the user should be able to decide but the choices are like those of a political election with only one candidate.
Over the last years, the notion of tooltips has somewhat changed.  
Let me quote the doc string of the tooltip class for the original definition:  

> Provides information about an item over which the user is hovering a cursor.
> The object should fire a show event when it appears.

MSDN gives a similar description, perhaps even narrower as they refer mainly to the mouse pointer.  
However, one has to read between the lines to establish a different view point.
In contrast to the classic definition, tooltips can now appear outside the app.  
In UIA terms, this would be "Peripheral", I think.  
Let's call this new type flyouts.
Some classic tooltips are quite useful, e.g. those that you get in an explorer window with modification date, size etc. for a file.
Taskbar tooltips that are announced whilst the focus isn't there are rather a nuisance IMHO.
It is perhaps difficult to implement a finer distinction in NVDA core (e.g. by adding flyouts to
the object presentation dialog).

I think the best we can do is adding filter capability to the tooltip class such that add-ons can profit from it.

This is the perfect occasion to use the extensionPoints.Filter class.
After all, it has been introduced for exactly that purpose, has it not.

**Comment by josephsl** [comment]
Hi,

Any additional comments?

The filter mechanism is useful for text processing, not really a way to let overlay classes change things here and there. To do that, one must be willing to add conditions in the class itself, or let global plugins and app modules take ownership of this class (inheritance and overriding).

Thanks.

**Comment by Robert-J-H** [comment]
I haven't the extensionPoints code at my fingertips right now but I
think there's also the Decider class.
Registered functions would be notified about the tooltip source and
the text to be spoken.
They can then return either True or False to allow or prevent the
speaking of it. Thx

**Comment by gregjozk** [comment]
Hello,

what is a status of this pr?

thx.

regards,
Jožef

**Comment by josephsl** [comment]
Hi,

Regarding deciders: this means extension points must be willing to deal with cases where tooltips are not announced because someone said "no" while the decider was running. I think this might not be what we want.

For now, I'd treat tooltips from UIA as just tooltips. If there are no major objections, I propose that this be ready for master branch soon.

Thanks.

**Comment by ehollig** [comment]
Hello, What is the status of this PR? It looks like it has been incubating for a while now.

**Comment by josephsl** [comment]
Hi, some small things here and there, and apart from that, and once it is approved, it should be ready for master branch. In case review comments are posted for this one, I’ll address them in November. Thanks.

**Comment by josephsl** [comment]
Hi,

Actually, it is ready to go - I think it's just the matter of when it'll be included in master branch, and I can address review comments in October.

Thanks.

**Comment by leonardder** [comment]
I removed the incubating label from this as we're no longer using that label. Having said that, the fact that this has been incubating for a while suggest that the code is of enough quality and has been approved in the past.

**Comment by josephsl** [comment]
Hi, just a few things here and there and should be ready for another round of reviews – at least I heard no major reports about this in Windows 10 App Essentials, so it could serve as a basis for a review next week. Thanks.

**Comment by josephsl** [comment]
Hi,

Let me merge in master first.

Thanks.

**Comment by josephsl** [comment]
Hi,

Regarding review bits: yes, approved, but it won't hurt to let another pair of eyes/ears/hands take a look at this one before making this out to master.

Thanks.

**Comment by Robert-J-H** [comment]
Hm, I still see a lot of issues (see my remarks above).
The tooltip behaviour is somewhat erratic (not necessarily related to
UIA alone). For instance:
The tool tip for the last focused task bar icon is often announced
between focus changes within an application. In general, this doesn't
produce anything valuable.
Tool tips for desktop icons with focus (e.g. shortcut target) are
often wrong as they can come from the mouse object, rather than the
focused object.
I see this behaviour also in Reaper.

**Comment by dingpengyu** [comment]
This PR has the following error in the build:
Failure: AppVeyor build failed
Hope that can be fixed
thank
cc
@josephsl

**Comment by josephsl** [comment]
Hi. Likely caused by the old code, which will be fixed once master branch is merged. Thanks.

**Comment by AmadeusW** [comment]
How can I test changes in this PR? @josephsl do you have files I can use to patch NVDA 2019.1.1? I can install any other version, too, if that would make it easier for you to share the fix.

I'd like to verify whether this addresses issue where XAML tooltips are not announced in Visual Studio (VS is a Win32 app using WPF\XAML, not an Universal Windows app)

**Comment by josephsl** [comment]
Hi, I have suspended development of this PR for now. Thanks.

**Comment by AmadeusW** [comment]
that's fine, I realized that NVDA does read some tooltips in VS, so this PR wouldn't have helped my case. Thank you!

**Comment by josephsl** [comment]
Hi,

I'll make sure this PR passes checks.

Thanks.

**Review by feerrenrut — APPROVED** [review]

**Comment by josephsl** [comment]
Hi, I’ll rebase it on master unless there is no conflict merging this into master. Thanks.

**Pull request merged**
